wordpress-mobile/gutenberg-mobile
# Scroll to caret position on focused field

**Fixes** https://github.com/wordpress-mobile/gutenberg-mobile/issues/436 https://github.com/wordpress-mobile/gutenberg-mobile/issues/432 

This PR solves some scrolling issues when keyboard opens. With this PR we are starting to make use of caret position to arrange the scroll position. 

Description of the new mechanism can be found in the [child PR](https://github.com/wordpress-mobile/react-native-keyboard-aware-scroll-view/pull/2).

Preview
![ezgif com-video-to-gif-6](https://user-images.githubusercontent.com/5032900/50592918-38654900-0ea7-11e9-95ab-8d6508e06ae9.gif)

**Implementation Details**
The position of the currently focused input in the window had been measured with a native call `UIManager.measureInWindow` previously. I've added a similar call `UIManager.measureSelectionInWindow` but this time it also fetched the caret position.
Scrolling to caret is only available for paragraph and heading blocks currently because it requires a TextView, currently Aztec's text input supports this. React native's text input doesn't provide us the caret position and hides the inner TextView from us. React native text inputs will continue to scroll to the bottom of the component. In the future we can replace those with Aztec's inputs to achieve same thing.

**Testing Instructions**

**Testing Prerequisites**

- Run `yarn install` and `yarn start`

To test with iOS example app 
- Run `yarn ios`

To test with WPiOS
- Follow instructions at [WPiOS PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/10730)
https://github.com/wordpress-mobile/gutenberg-mobile/issues/436 needs to be tested with WPiOS

Test 1 - Test with long text
This tests https://github.com/wordpress-mobile/gutenberg-mobile/issues/432
- Add a paragraph block and add a text that is longer than the viewable area when keyboard is open
- Close keyboard if open
- Align top of the paragraph block to top of the screen
- Tap at the first line
- Verify that no scrolling occurs because caret is already in the viewable area
- Change orientation and verify caret is still visible

Test 2 - Test with the last line
- Use a paragraph block with some multiple line of input
- Close keyboard if open
- Align the block to the bottom of the screen
- Tap to the last line
- Verify that scroll position is adjusted and inner toolbar is also visible
- Change orientation and verify caret is still visible

Test 3 - Test with a line other than the last line
- Use a paragraph block with some multiple line of input
- Close keyboard if open
- Align the block to the bottom of the screen
- Tap to some line other than the last line
- Verify that scroll position is adjusted to make caret visible only
- Change orientation and verify caret is still visible

Test devices should include:
- iPhone X variety device
- iPad 
- Either one of iPhone 6-7-8

Test 4 - Android
- Android side wasn't affected so a simple test would be enough
Open Android example app and verify it opens successfully
